xds/internal/balancer/clusterresolver/resource_resolver_dns.go

```diff
@@ -83,16 +90,33 @@ func newDNSResolver(target string, topLevelResolver topLevelResolver, logger *gr
 		return ret
 	}
 
-	r, err := newDNS(resolver.Target{URL: *u}, ret, resolver.BuildOptions{})
-	if err != nil {
+	// Create the serializer and schedule a callback to create the actual DNS
+	// resolver. This needs to happen asynchornously to avoid the following
+	// deadlock:
+	// - this method is called from rr.updateMechanisms which holds `rr.mu`
+	// - dns resolver Build() can report state or error inline, which is handled
+	//   in UpdateState or ReportError, both of which call
+	//   topLevelResolver.onUpdate, which tries to grab `rr.mu`, and hence
+	//   deadlocks
+	ctx, cancel := context.WithCancel(context.Background())
+	ret.serializer = grpcsync.NewCallbackSerializer(ctx)
+	ret.serializerCancel = cancel
+	ret.serializer.Schedule(func(context.Context) {
+		r, err := newDNS(resolver.Target{URL: *u}, ret, resolver.BuildOptions{})
+		if err == nil {
+			ret.dnsR = r
+			return
+		}
+
 		if ret.logger.V(2) {
 			ret.logger.Infof("Failed to build DNS resolver for target %q: %v", target, err)
 		}
+		ret.mu.Lock()
 		ret.updateReceived = true
+		ret.mu.Unlock()
 		ret.topLevelResolver.onUpdate()
```

Comment:
This seems correct thinking about it. My musing is super minor here. Previously, when you called this function and an onUpdate() gets called, it's considered having "received" configuration for this particular discovery mechanism and will trigger fallback if DNS Resolver hasn't sent anything yet. Now, it happens async (and we wait for the dns resolver to build before to set the bool). Oh, I guess we can't consider this discovery mechanism having had a chance to received an update before the dns resolver has a chance to return results inline. I was mainly concerned now there's a new time frame between when we build this resource resolver and when this callback actually executes where previously before it would build the config and trigger fallback (in the case of no update sent so no addrs), and now it just waits until onUpdate() is called here. But this wait seems minor and ok. So seems correct.

Reply:
Thanks for this comment. I realized that I was not handling all cases correctly. Specifically, if a url parse failure happened, with my previous commit, things would still deadlock.

So, I thought about this for a while, and came to the conclusion that the correct place to handle this is in the `resourceResolver` component. So, my current commit handles this:
- it makes the `onUpdate` non blocking, but pushing the call to `generateLocked` via a callback serializer
- make it possible for the `cluster_resolver` LB policy to notify the `resourceResolver` whether it is being actually closed or not (in `stop()`)

The use of the callback serializer is *not* absolutely required in the `resourceResolver`. I could as well have pushed a signal on an unbounded buffer, and have a goroutine read from it. But using a callback serializer just made is easier. Let me know if you have any questions/concerns on this approach.

Reply:
Oh yeah, I see now about url parse error triggering deadlock with onUpdate(). There was a discussion about adding an nack validation in the xDS Client in the xDS Chat, but the discussion got tabled. Then there would no longer need to be this error handling. Let me think some about what layer you put the buffer of callbacks at.